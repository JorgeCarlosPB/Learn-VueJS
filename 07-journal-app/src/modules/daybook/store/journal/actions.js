
//Actions son acciones asincronas para llamar mutaciÃ³n
// export const myAction = async ({commit}) =>{

import journalApi from "@/api/journalApi"

    
// }

export const loadEntries = async ({commit}) =>{
    const {data} = await journalApi.get('/entries.json')
    if( !data){
        commit('setEntries', [])
        return
    } 
    const entries = []
    for (let id of Object.keys(data)){
        entries.push({
            id,
            ...data[id]
        })
    }

    commit('setEntries', entries)
}

export const updateEntry = async ({commit}, entry) => {                //entry debe de ser un parametro

    const {date, picture, text} = entry
    const dataToSave = {date, picture, text}

    const resp=await journalApi.put(`/entries/${entry.id}.json`,dataToSave)
    dataToSave.id = entry.id
    //Extraer solo lo que se necesita  //-id

    //await jounrlaApi.put(PATH .json, dataToSave)

    //Commit de una mutation -> updateEntry
    commit('updateEntry',{...dataToSave})

}

export const createEntry = async ({commit}, entry) => {
    //dataToSave
    const {date,picture, text} = entry
    const dataToSave = {date,picture, text}

    const {data} = await journalApi.post(`entries.json`,dataToSave)
    //cons {data} = await journalApi.post(PATH.json, dataToSave)

    dataToSave.id = data.name
    commit('addEntry', dataToSave)
    //data = {"name": "MarijsKGESa-SGsef"}

    return data.name
}

export const deleteEntry = async ({commit},id)=>{

    //Await jorunalApi.delete(path)
    await journalApi.delete(`/entries/${id}.json`)

    commit('deleteEntry',id)

    return id

}