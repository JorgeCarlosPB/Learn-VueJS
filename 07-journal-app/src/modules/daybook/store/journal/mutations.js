//son sincronas y hacen modificaciones del state
// export const myAction = {state} =>{
    
// }

export const setEntries = (state, entries) =>{
    state.entries = [...state.entries, ...entries]
    state.isLoading = false
    
}

export const updateEntry = (state, entry) =>{  //entry actualizada

    //state.entries => un arreglo...
    const idx = state.entries.map(e => e.id).indexOf(entry.id)
    state.entries[idx]=entry

    //state.entries => 
    
}

export const addEntry = (state, entry) =>{
    //state ->entries -> la nueva entrada debe de ser la primera

    state.entries = [entry, ...state.entries]
}

export const deleteEntry = (state,id) => {
    //remover del state.entries borrar la entrada por ese ID

    state.entries = state.entries.filter(entry => entry.id !== id)
}