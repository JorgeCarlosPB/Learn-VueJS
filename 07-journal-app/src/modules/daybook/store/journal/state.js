
//el state es reactivo y cuando cambie se notifica a todos los lugres que estÃ¨n escucando el state
// export default () =>({
    
// })

export default()=>({
    isLoading: true,
    entries: []

})