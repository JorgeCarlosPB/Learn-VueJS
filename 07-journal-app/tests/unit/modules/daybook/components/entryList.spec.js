
import { shallowMount } from '@vue/test-utils'
import {createStore} from 'vuex'

import journal from '@/modules/daybook/store/journal'

import EntryList from '@/modules/daybook/components/EntryList.vue'
import {journalState} from '../../../mock-data/test-journal-state'

const createVuexStore = (initialState) =>
    createStore({
        modules: {
            journal: {
                ...journal,
                state: {...initialState}
            }
        }
    })


describe('Pruebas en el entryList',()=>{

    /*const journalMockModule = {
        namespaced: true,
        getters: {
            //getEntriresbyTerm: jest.fn()
            getEntriesByTerm
        },
        state: ()=>({
            isLoading: false,
            entries: journalState.entries
        })

    }

    const store = createStore({
        modules: {
            journal: {...journalMockModule}
        }
    })*/


    const store = createVuexStore(journalState)
    const mockRouter = {
        push: jest.fn()
    }

    let wrapper 
    beforeEach(()=>{
        jest.clearAllMocks()
        wrapper = shallowMount(EntryList, {
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })
     

    test('Debe de llamar el getEntriesByTerm sin tÃ©rmino y mostrar 2 entradas', ()=>{

        console.log(wrapper.html())

        expect(wrapper.findAll('my-entry-stub').length).toBe(2)
        expect(wrapper.html()).toMatchSnapshot()

    })

    test('Debe de llamar el getEntriesByTerm y filtrar las entradas', async()=>{
        const input = wrapper.find('input')
        await input.setValue('segunda')

        expect(wrapper.findAll('my-entry-stub').length).toBe(1)
        //expect(wrapper.html()).toMatchSnapshot()

    })

    test('el boton debe de redireccionar a /new', ()=>{
        wrapper.find('button').trigger('click')
        
        expect(mockRouter.push)
            .toHaveBeenCalledWith({name: 'entry', params: {id: 'new'}})

        //expect(wrapper.findAll('entry-stub').length).toBe(1)
        //expect(wrapper.html()).toMatchSnapshot()

    })

    
})