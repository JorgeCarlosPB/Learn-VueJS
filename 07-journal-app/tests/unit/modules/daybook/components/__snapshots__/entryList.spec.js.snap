// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el entryList Debe de llamar el getEntriesByTerm sin t√©rmino y mostrar 2 entradas 1`] = `
<div class="entry-list-container">
  <div class="px-2 pt-2"><input type="text" class="form-control" placeholder="Buscar entrada"></div>
  <div class="mt-2 d-flex flex-column"><button class="btn btn-primary mx-3"><i class="fa fa-plus-circle"></i> Nueva entrada </button></div>
  <div class="entry-scrollarea">
    <my-entry-stub entry="[object Object]"></my-entry-stub>
    <my-entry-stub entry="[object Object]"></my-entry-stub>
  </div>
</div>
`;
